tests/test_bootstrap.py

```diff
@@ -0,0 +1,551 @@
+        """
+        # prepare bootstrap and distribution
+        bs = Bootstrap().get_bootstrap("sdl2", self.ctx)
+        self.setUp_distribution_with_bootstrap(bs)
+
+        expected_bootstraps = {"empty", "service_only", "webview", "sdl2"}
+        set_of_bootstraps = set(Bootstrap().list_bootstraps())
+        self.assertEqual(
+            expected_bootstraps, expected_bootstraps & set_of_bootstraps
+        )
+        self.assertEqual(len(expected_bootstraps), len(set_of_bootstraps))
+
+    def test_get_bootstraps_from_recipes(self):
+        """A test which will initialize a bootstrap and will check if the
+        method :meth:`~pythonforandroid.bootstrap.Bootstrap.
+        get_bootstraps_from_recipes` returns the expected values
+        """
+        recipes_service = {"pyjnius", "python3"}
+        bs = Bootstrap().get_bootstrap_from_recipes(recipes_service, self.ctx)
+        self.setUp_distribution_with_bootstrap(bs)
+
+        self.assertEqual(bs.name, "service_only")
+
+        # test wrong recipes
+        wrong_recipes = {"python2", "python3", "pyjnius"}
+        bs = Bootstrap().get_bootstrap_from_recipes(wrong_recipes, self.ctx)
+        self.assertIsNone(bs)
+
+    @mock.patch("pythonforandroid.bootstrap.ensure_dir")
+    def test_prepare_dist_dir(self, mock_ensure_dir):
+        """A test which will initialize a bootstrap and will check if the
+        method :meth:`~pythonforandroid.bootstrap.Bootstrap.prepare_dist_dir`
+        successfully calls once the method `endure_dir`
+        """
+        # prepare bootstrap and distribution
+        bs = Bootstrap().get_bootstrap("sdl2", self.ctx)
+        self.setUp_distribution_with_bootstrap(bs)
+
+        bs.prepare_dist_dir("fake_name")
+        mock_ensure_dir.assert_called_once_with(bs.dist_dir)
+
+    @mock.patch("pythonforandroid.bootstrap.open", create=True)
+    @mock.patch("pythonforandroid.util.chdir")
+    @mock.patch("pythonforandroid.bootstrap.shutil.copy")
+    @mock.patch("pythonforandroid.bootstrap.os.makedirs")
+    @mock.patch("pythonforandroid.bootstrap.sh.ln")
+    @mock.patch("pythonforandroid.bootstrap.listdir")
+    @mock.patch("pythonforandroid.bootstrap.sh.mkdir")
+    @mock.patch("pythonforandroid.bootstrap.sh.rm")
+    @mock.patch("pythonforandroid.bootstrap.isdir")
+    def test_bootstrap_prepare_build_dir(
+        self,
+        mock_isdir,
+        mock_sh_rm,
+        mock_sh_mkdir,
+        mock_listdir,
+        mock_sh_ln,
+        mock_os_makedirs,
+        mock_shutil_copy,
+        mock_chdir,
+        mock_open,
+    ):
+        """A test which will initialize a bootstrap and will check if the
+        method :meth:`~pythonforandroid.bootstrap.Bootstrap.prepare_build_dir`
+        successfully calls the methods that we need to prepare a build dir.
+        """
+        mock_isdir.return_value = True
+        mock_sh_mkdir.return_value = "sh.mkdir successfully mocked"
+        mock_sh_rm.return_value = "sh.rm successfully mocked"
```

Comment:
Minor: just curious why would we need to set the return values for some of these like `mock_sh_rm` and `mock_sh_mkdir`?

Reply:
You are right, We don't need to do this, Thanks @AndreMiras, I forgot to remove it :smile: